Refinement of bound vars using guard BIFs for mismatching clauses

zuiderkwast opened the pull request:
This now passes:

```Erlang
-spec refine_bound_var_by_guard_bifs(ok | integer() | pid() |
                                     #{x => y} | {x, y}) -> ok.
refine_bound_var_by_guard_bifs(X) ->
    if
        is_map(X)     -> ok;
        is_tuple(X)   -> ok;
        is_integer(X) -> ok;
        is_pid(X)     -> ok;
        true          -> X
    end.
```

Fixes some of the examples in #25.

FrankBro commented on src/typechecker.erl:
```diff
@@ -3423,6 +3431,47 @@ refinable(_) ->
+            end;
+        _OtherGuards ->
+            %% No refinement
+            VEnv
```
Great work. Would it be hard to add `andalso`/`and` and `orelse`/`or`?

zuiderkwast replied:
Possibly for in limited cases, but I'd prefer to leave it to a separate PR.

For a guard sequence like `is_atom(X); is_integer(X)`, where all the guards are on the same variable X, the type of X can be refined by excluding both atom() and integer().

Do you have any other example in mind?

FrankBro replied:
Totally fine to split it in multiple PRs.

This is a hair-pulling example a bit. Only useful if we want to have it as a test I guess.

```erlang
-spec test(integer() | atom() | list()) -> list().
test(X) ->
    if
        is_atom(X) orelse is_integer(X) -> [];
        true -> X
    end.
```

It also reminded me of the record part I did. Could also be another PR.

```erlang
-record(r1, {f :: integer()}).
-record(r2, {f :: integer()}).

-spec record(#r1{} | #r2{}) -> integer().
record(#r1{f = F}) -> F;
record(R2) -> R2#r2.f.
```

zuiderkwast commented:
> My only concern is that the rules for when refinement happens become increasingly complex. Please update the wiki so we can have a good sense of how we communicate to the programmer when they can rely on refinement.

I agree. @FrankBro can you update with what's already been merged? Then I can add stuff from this PR. It's not good to let documentation lag behind.

zuiderkwast commented:
Documentation draft:

### Refinement of bound variables after mismatching clauses

The type of a variable `Var` bound already before an `if`, `case` or other construct with clauses, can be refined using information about why the previous clauses have failed to match. The following conditions must be fulfilled for this refinement to happen:

1. The patterns of a clause must match all possible input (meaning that every pattern is either `_` or a free variable)
    * Note: `if` clauses have no patterns, so this condition is automatically fulfilled
2. There is a single guard on the form `is_TYPE(Var)` for the clause and no other guards are present

If these conditions are fulfilled, the only reason for the clause to mismatch is the guard. Thus, for the following clauses within the same construct (`if` or `case` etc.), the type of Var can be refinement by *excluding* TYPE from the type of Var.

Example:

```Erlang
-spec f(integer() | atom()) -> atom().
f(Var) ->
    case Var of
        _ when is_integer(Var) -> hello;
        _                      -> Var       % here Var :: atom()
    end.
```

FrankBro commented:
Yes, will try to come up with something asap.

FrankBro commented:
Can we have a copy of the wiki in a folder somewhere so we can do PRs on it? I feel like this PRs changes to the wiki and the one for guards is kind of related.

Zalastax commented:
@FrankBro, GitHub doesn't provide an easy way to use that workflow. There are workarounds though: https://www.growingwiththeweb.com/2016/07/enabling-pull-requests-on-github-wikis.html. It's okay if you just make the edits you want to do. Someone can make additional changes afterwards if they have any comments.

FrankBro commented:
Which is why I was suggesting having a copy in like doc or something.

I tried to merge all three PRs, idk if it's good. I suck at documentation 😭 

### Refinement of bound variables
The type of a variable Var bound already before an if, case or other construct with clauses, can be refined using information such as guards, record matching as well as why the previous clauses have failed to match. The following conditions must be fulfilled for this refinement to happen:

The patterns of a clause must match all possible input (meaning that every pattern is either _ or a free variable)
Note: if clauses have no patterns, so this condition is automatically fulfilled
There is a single guard on the form is_TYPE(Var) for the clause and no other guards are present
If these conditions are fulfilled, we can refine the variable. Additionally, if the only reason for the clause to mismatch is the guard, we can refine the type for the following clauses. Thus, for the following clauses within the same construct (if or case etc.), the type of Var can be refinement by excluding TYPE from the type of Var.

Example:
```erlang
-record(r, {f :: atom()}).
-spec f(integer() | atom() | #r{}) -> atom().
f(Var) ->
    case Var of
        R = #r{}                         -> R#r.f;
        _ when is_integer(Var) -> hello;
        _                                     -> Var       % here Var :: atom()
    end.
```

zuiderkwast commented:
@FrankBro, the clause `R = #r{} -> R#r.f` is not about refining *already bound variables* for *mismatching* clauses. R is a variable *bound in the clause* if the clause is *matching*. Thus, I don't think it belongs in the same section.

zuiderkwast commented:
I updated the [wiki page](https://github.com/josefs/Gradualizer/wiki/Type-refinement) and added stubs for the matching clauses as well. The two main sections are for *matching clauses* and for *mismatching clauses*.

Feel free to add stuff!

Another possibility for work-in-progress documentation is to create a temporary wiki page for the draft and/or discussions. (Compare to Wikipedia's discussion tab of any page. It's just another wiki page.)
